File: windows/src/ChatUnpack.Core/Export/MarkdownChunker.cs
using System.Globalization;

namespace ChatUnpack.Core.Export;

public sealed class MarkdownChunker
{
  private const int MarkerReserve = 80;

  public MarkdownChunker(int maximumCharacters = 1800)
  {
    if (maximumCharacters < 200)
    {
      throw new ArgumentOutOfRangeException(nameof(maximumCharacters));
    }

    MaximumCharacters = maximumCharacters;
  }

  public int MaximumCharacters { get; }

  public List<string> Split(string markdown)
  {
    var value = markdown ?? string.Empty;
    if (TextElementCount(value) <= MaximumCharacters)
    {
      return [value];
    }

    var payloadLimit = MaximumCharacters - MarkerReserve;
    var payloads = SplitPayload(value, payloadLimit);
    var total = payloads.Count;
    return payloads
      .Select((payload, index) => $"{Marker(index + 1, total)}\n\n{payload}")
      .ToList();
  }

  private List<string> SplitPayload(string markdown, int limit)
  {
    var remaining = markdown;
    var payloads = new List<string>();
    while (TextElementCount(remaining) > limit)
    {
      var prefix = TakeTextElements(remaining, limit);
      var splitIndex = PreferredSplitIndex(prefix, limit / 2)
        ?? prefix.Length;
      payloads.Add(remaining[..splitIndex]);
      remaining = remaining[splitIndex..];
    }

    if (remaining.Length > 0)
    {
      payloads.Add(remaining);
    }

    return payloads;
  }

  private static int? PreferredSplitIndex(string prefix, int minimumOffset)
  {
    var separators = new[] { "\n\n---\n\n", "\n\n", "\n" };
    foreach (var separator in separators)
    {
      var index = prefix.LastIndexOf(separator, StringComparison.Ordinal);
      if (index < 0)
      {
        continue;
      }

      var end = index + separator.Length;
      if (TextElementCount(prefix[..end]) >= minimumOffset)
      {
        return end;
      }
    }

    return null;
  }

  private static string TakeTextElements(string value, int count)
  {
    var starts = StringInfo.ParseCombiningCharacters(value);
    if (starts.Length <= count)
    {
      return value;
    }

    return value[..starts[count]];
  }

  private static int TextElementCount(string value)
  {
    return StringInfo.ParseCombiningCharacters(value).Length;
  }

  private static string Marker(int part, int total)
  {
    return part == total
      ? $"【聊天记录分段 {part}/{total}，已发送完毕，请统一处理全部分段】"
      : $"【聊天记录分段 {part}/{total}，请等待全部 {total} 段发送完成后统一处理】";
  }
}

File: Sources/ChatUnpackCore/Export/MarkdownChunker.swift
public struct MarkdownChunker: Sendable {
  public let maximumCharacters: Int

  private let markerReserve = 80

  public init(maximumCharacters: Int = 1800) {
    precondition(maximumCharacters >= 200)
    self.maximumCharacters = maximumCharacters
  }

  public func split(_ markdown: String) -> [String] {
    guard markdown.count > maximumCharacters else { return [markdown] }

    let payloadLimit = maximumCharacters - markerReserve
    let payloads = splitPayload(markdown, limit: payloadLimit)
    let total = payloads.count

    return payloads.enumerated().map { index, payload in
      "\(marker(part: index + 1, total: total))\n\n\(payload)"
    }
  }

  private func splitPayload(_ markdown: String, limit: Int) -> [String] {
    var remaining = markdown
    var payloads: [String] = []

    while remaining.count > limit {
      let limitIndex = remaining.index(remaining.startIndex, offsetBy: limit)
      let prefix = remaining[..<limitIndex]
      let splitIndex = preferredSplitIndex(in: prefix, minimumOffset: limit / 2)
        ?? limitIndex
      payloads.append(String(remaining[..<splitIndex]))
      remaining = String(remaining[splitIndex...])
    }

    if !remaining.isEmpty {
      payloads.append(remaining)
    }
    return payloads
  }

  private func preferredSplitIndex(
    in prefix: Substring,
    minimumOffset: Int
  ) -> String.Index? {
    for separator in ["\n\n---\n\n", "\n\n", "\n"] {
      guard let range = prefix.range(of: separator, options: .backwards) else { continue }
      let offset = prefix.distance(from: prefix.startIndex, to: range.upperBound)
      if offset >= minimumOffset {
        return range.upperBound
      }
    }
    return nil
  }

  private func marker(part: Int, total: Int) -> String {
    if part == total {
      return "【聊天记录分段 \(part)/\(total)，已发送完毕，请统一处理全部分段】"
    }
    return "【聊天记录分段 \(part)/\(total)，请等待全部 \(total) 段发送完成后统一处理】"
  }
}
